Documentation/planning/multi-sdk-band-support.md

```diff
@@ -0,0 +1,253 @@
+- Building N SDK bands **using**
+- The *outputs* of the runtime build combined with additional previously source-built artifacts (e.g. a matching band SDK, required tools and libraries) and source, to produce N SDK bands that meet source-build requirements.
+
+The runtime build requires a 1xx band SDK to build **and** redistributes functionality from tooling that ships in the 1xx band (e.g. it has dependencies on roslyn and roslyn-analyzers). Furthermore, most changes to a .NET major version occur when only one SDK band exists. Therefore it makes practical sense at this point to include the source of the 1xx SDK and the shared runtime in the same VMR branch, and to produce all artifacts related to 1xx, including the runtime, in its build. Newer bands will be VMRs that contain **only** the components that **differ** from the 1xx band. Only those components that differ. When building the newer band, the outputs of the 1xx build would be fed into the newer band build, in a similar manner as previously source built artifacts are today. If the source for a component does not exist in a given VMR, it is simply not built, and obviously no outputs would be produced for that component. Thus any downstream dependencies would be forced to use the versions provided by input source built artifacts, typically from the 1xx source build. If no such version exists, then the input would fail to be found or restored from online sources and be reported as a prebuilt.
+
+This approach fits naturally into the source-build methodology. Generally, additional SDK bands may be referred to as "subsetted VMRs", since use of a subsetted VMR is not restricted to SDKs only.
+
+### Conceptual splitting of 'previously source-built' artifact types
+
+Today, the term "previously source-built artifacts' refers to those artifacts that were built in a previous product build iteration. Typically N-1. For example, they would refer to the outputs of 8.0.1's build, when building 8.0.2. When building a subsetted VMR, two sets of input sets of source-built artifacts are required. While similar in some ways, they should be kept distinct for clarity. They are referred to as:
+
+- **Previously source-built (PSB)** - Artifacts from a previous servicing/preview iteration, or from a bootstrap build. These may be used when building the current VMR, but may not be bundled into any output.
+- **Current source-built (CSB)** - Artifacts just built, e.g. from a 1xx VMR build. These may be used **and** bundled into the build outputs, since logically they are from the same source-build iteration.
+
+### Supported input artifact combinations and validation thereof
+
+Not all toolsets can be used to build all VMRs. For instance, it is unlikely to be the case that the 1xx SDK will successfully build all 2xx+ bands. Components of newer SDKs frequently take dependencies on newer SDK features. However, the following statements can be made:
```

Comment:
I understand why this uncertainty exists, but I'm curious why we don't have more strict rules on dependencies of SDK components on SDK features that would make this behavior predictable. What would be the implications of requesting the product to be always buildable with the previous SDK bands?

Reply:
The SDK bands are driven by VS development, which moves forward between .NET releases. Those components of the .NET SDK need to introduce new functionality, which may require support from the SDK.

The implication of requiring the SDK bands to be always buildable with earlier bands is that you couldn't introduce some new features. In addition, it wouldn't improve distro maintainer workflow by much. You're still building the 1xx and Nxx bands. Whether you depend on the Nxx band to build the Nxx band or not is only a tiny difference.

Reply:
> The implication of requiring the SDK bands to be always buildable with earlier bands is that you couldn't introduce some new features.

I'm still not comfortable with this statement. In my thinking, we are mixing together two things - introducing an SDK feature (that is later used by the corresponding VS version) and using this feature to build the SDK itself. It's clear the compiler and other tooling want to dogfood the latest features they introduce aggressively, but I still don't see a case where it wouldn't be possible to introduce a set of new features solely through the features available in existing version of the SDK.

Reply:
I agree with @tkapin. In an ideal world, every feature band should be buildable using 1xx. But at minimum, it should be buildable using the previous band. Anything else breaks N+1. 

An ideal packaging approach for me would be to have a package of dotnet that builds 1xx, and a package of dotnet that builds the latest feature band, using the artifacts of 1xx. Considering that this isn't on the table, I can imagine then packaging 2xx sdk to build 3xx sdk once that's released.

Reply:
I don't think we can make that guarantee. I think it may work in most cases, but in others, a bootstrap of the new band will be required. Even if we get an agreement from VS that they would not break this workflow, new functionality is being introduced in those bands, some of which may require SDK support, and so bugs are possible.

In conclusion,

**We can _always_ say (barring catastrophic problems), that 200 GA will build 201, will build 202, etc.** This is the same garuantee we make with the 1xx band+runtime today. It just may be that a 200 bootstrap will be required to build the initial 200.

Reply:
I clarified the logic here. Can you take a look once I push the latest update?